src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -126,11 +127,36 @@ public RecoverySourceHandler(final IndexShard shard, final StartRecoveryRequest
         this.response = new RecoveryResponse();
     }
 
-    /**
-     * @return the {@link RecoveryResponse} after the recovery has completed all three phases
-     */
-    public RecoveryResponse getResponse() {
-        return this.response;
+    /** performs the recovery from the local engine to the target */
+    public RecoveryResponse recoverToTarget() {
+        try (Translog.View translogView = shard.translog().view()) {
+            logger.trace("captured translog id [{}] for recovery", translogView.minTrasnlogId());
+            final SnapshotIndexCommit phase1Snapshot;
+            try {
+                phase1Snapshot = shard.snapshotIndex(false);
+            } catch (Throwable e) {
+                Releasables.closeWhileHandlingException(translogView);
+                throw new RecoveryEngineException(shard.shardId(), 1, "Snapshot failed", e);
+            }
+
+            try {
+                phase1(phase1Snapshot, translogView);
+            } catch (Throwable e) {
+                throw new RecoveryEngineException(shard.shardId(), 1, "phase1 failed", e);
+            } finally {
+                Releasables.closeWhileHandlingException(phase1Snapshot);
+            }
+
+            logger.trace("snapshot translog for recovery. current size is [{}]", translogView.estimatedTotalOperations());
+            try (Translog.Snapshot phase2Snapshot = translogView.snapshot()) {
+                phase2(phase2Snapshot);
+            } catch (Throwable e) {
+                throw new RecoveryEngineException(shard.shardId(), 2, "phase3 failed", e);
```

Comment:
I think this should be "phase2 failed" instead of "phase3 failed" since we only have 2 phases now?